src/lib.rs

```diff
@@ -116,40 +117,60 @@ pub struct Rules {
 }
 
 impl Rules {
-    /// Split a text into segments.
-    pub fn split<'a>(&self, text: &'a str) -> Vec<&'a str> {
+    /// Obtain the ranges for text segments. Guaranteed to be at character bounds.
+    pub fn split_ranges(&self, text: &str) -> Vec<Range<usize>> {
         let mut segments = Vec::new();
 
-        let mut mask: Vec<Option<bool>> = vec![None; text.len()];
+        // Mask defines the segments of text which are
+        // TODO use a proper tri-state enum here
+        let mut masked_bytes: Vec<Option<bool>> = vec![None; text.len()];
 
-        for rule in &self.rules {
-            for index in rule.match_indices(text) {
-                if index >= text.len() {
-                    continue;
+        'outer: for rule in &self.rules {
+            for byte_range in rule.match_indices(text) {
+                // byte range of non-content chars
+                let Range { start, end } = byte_range;
+
+                debug_assert!(end <= text.len());
+                if start >= text.len() {
+                    continue 'outer;
                 }
 
-                if mask[index].is_none() {
-                    mask[index] = Some(rule.do_break());
+                let byte_index = byte_range.start;
+                if masked_bytes[byte_index].is_none() {
+                    masked_bytes[byte_index] = Some(rule.do_break());
                 }
             }
         }
 
-        let mut prev_index = 0;
+        let mut prev_byte_pos = 0;
 
-        for (i, mask_val) in (0..text.len()).zip(mask) {
-            if let Some(true) = mask_val {
-                segments.push(&text[prev_index..i]);
-                prev_index = i;
+        // Iterate over characters, we don't want no half characters in the output ranges
+        for ((byte_pos, _c), tristate_mask) in text.char_indices().zip(masked_bytes) {
```

Comment:
You're zipping an iterator over chars with the byte-level mask here, right?. I'm pretty sure this will fail for multibyte characters. Could you fix this and add a test for this?

Reply:
That indeed is a bug.